README.md
# Bigstringaf_simd

A fork of bigstringaf_simd with support for SIMD intrinsics

# License
Please refer to LICENSE for original license. All added files will
have their respective license


CHANGES.md
### 0.4.0 (2018-10-26)

* freestanding: fix dependencies (ocaml-freestanding #22 @hannesm)
* xen: unify usage of mirage-xen-posix package (#17 @hannesm)
* fix typo in comment (#18 @tiensonqin)
* jbuild: do not use bash (#16 #21 @rgrinberg, #22 @hannesm)
* opam: add '"-p" name' to subst command (#15 @seliopou)

### 0.3.0 (2018-07-07)

* Add linking support for mirage-xen-ocaml and ocaml-freestanding (#12, #13, #14, h/t @samoht, @hannesm, @dinosaure)

### 0.2.0 (2018-06-10)

* Add memcmp operations (#2)
* Fix bounds checking bugs in constructors (#4, #5, h/t @yallop)
* Add safe blit/memcmp operations (#8)

### 0.1.0 (2018-04-01)

* initial release


lib/bigstringaf_simd.c
#include <caml/mlvalues.h>
#include <caml/memory.h>
#include <caml/fail.h>
#include <caml/custom.h>
#include <caml/bigarray.h>

#include <stdlib.h>
#include <stdint.h>

void bigstringaf_simd_finalize(value v);

static struct custom_operations bigstringaf_simd_ops = {
  "bigstringaf_simd_ops",
  bigstringaf_simd_finalize,
  custom_compare_default,
  custom_hash_default,
  custom_serialize_default,
  custom_deserialize_default,
  custom_compare_ext_default,
  custom_fixed_length_default
};

int bigstringaf_simd_element_size[] =
{ 4 /*FLOAT32*/, 8 /*FLOAT64*/,
  1 /*SINT8*/, 1 /*UINT8*/,
  2 /*SINT16*/, 2 /*UINT16*/,
  4 /*INT32*/, 8 /*INT64*/,
  sizeof(value) /*CAML_INT*/, sizeof(value) /*NATIVE_INT*/,
  8 /*COMPLEX32*/, 16 /*COMPLEX64*/,
  1 /*CHAR*/
};

value bigstringaf_simd_alloc(int flags, int num_dims, void* data, intnat* dim) {
  uintnat num_elts, asize, size;
  int i;
  value res;
  struct caml_ba_array * b;
  intnat dimcopy[CAML_BA_MAX_NUM_DIMS];

  CAMLassert(num_dims >= 0 && num_dims <= CAML_BA_MAX_NUM_DIMS);
  CAMLassert((flags & CAML_BA_KIND_MASK) <= CAML_BA_CHAR);
  for (i = 0; i < num_dims; i++) dimcopy[i] = dim[i];
  size = 0;
  if (data == NULL) {
    num_elts = 1;
    for (i = 0; i < num_dims; i++) {
      if (caml_umul_overflow(num_elts, dimcopy[i], &num_elts))
        caml_raise_out_of_memory();
    }
    if (caml_umul_overflow(num_elts,
                           bigstringaf_simd_element_size[flags & CAML_BA_KIND_MASK],
                           &size))
      caml_raise_out_of_memory();
    int ret = posix_memalign(&data, 32, size);
    if (((data == NULL) && (size != 0)) || ret == 12) caml_raise_out_of_memory();
    flags |= CAML_BA_MANAGED;
  }
  asize = SIZEOF_BA_ARRAY + num_dims * sizeof(intnat);
  res = caml_alloc_custom_mem(&bigstringaf_simd_ops, asize, size);
  b = Caml_ba_array_val(res);
  b->data = data;
  b->num_dims = num_dims;
  b->flags = flags;
  b->proxy = NULL;
  for (i = 0; i < num_dims; i++) b->dim[i] = dimcopy[i];
  return res;
}

value bigstringaf_simd_create(value vkind, value vlayout, value vdim)
{
  intnat dim[CAML_BA_MAX_NUM_DIMS];
  mlsize_t num_dims;
  unsigned int i;
  int flags;

  num_dims = Wosize_val(vdim);
  /* here num_dims is unsigned (mlsize_t) so no need to check (num_dims >= 0) */
  if (num_dims > CAML_BA_MAX_NUM_DIMS)
    caml_invalid_argument("Bigarray.create: bad number of dimensions");
  for (i = 0; i < num_dims; i++) {
    dim[i] = Long_val(Field(vdim, i));
    if (dim[i] < 0)
      caml_invalid_argument("Bigarray.create: negative dimension");
  }
  flags = Caml_ba_kind_val(vkind) | Caml_ba_layout_val(vlayout);
  return bigstringaf_simd_alloc(flags, num_dims, NULL, dim);
}

void bigstringaf_simd_finalize(value v)
{
  struct caml_ba_array * b = Caml_ba_array_val(v);

  switch (b->flags & CAML_BA_MANAGED_MASK) {
  case CAML_BA_EXTERNAL:
    break;
  case CAML_BA_MANAGED:
    if (b->proxy == NULL) {
      free(b->data);
    } else {
      if (-- b->proxy->refcount == 0) {
        free(b->proxy->data);
        free(b->proxy);
      }
    }
    break;
  case CAML_BA_MAPPED_FILE:
    /* Bigarrays for mapped files use a different finalization method */
    /* fallthrough */
  default:
    CAMLassert(0);
  }
}


lib/bigstringaf_simd_avx2.c
#include <caml/mlvalues.h>
#include <caml/bigarray.h>
#include <immintrin.h>
#include <stdint.h>

#define get_ops1(a) \
  uint8_t* op_a = Caml_ba_data_val(Field(a,0)); \
  size_t   o1   = Long_val(Field(a,1))

#define get_ops2(a,b) \
  uint8_t* op_a = Caml_ba_data_val(Field(a,0)); \
  uint8_t* op_b = Caml_ba_data_val(Field(b,0)); \
  size_t   o1   = Long_val(Field(a,1)); \
  size_t   o2   = Long_val(Field(b,1))

#define get_ops3(a,b,c) \
  uint8_t* op_a = Caml_ba_data_val(Field(a,0)); \
  uint8_t* op_b = Caml_ba_data_val(Field(b,0)); \
  uint8_t* op_c = Caml_ba_data_val(Field(c,0)); \
  size_t   o1   = Long_val(Field(a,1)); \
  size_t   o2   = Long_val(Field(b,1)); \
  size_t   o3   = Long_val(Field(c,1))

#define get_ops4(a,b,c,d) \
  uint8_t* op_a = Caml_ba_data_val(Field(a,0)); \
  uint8_t* op_b = Caml_ba_data_val(Field(b,0)); \
  uint8_t* op_c = Caml_ba_data_val(Field(c,0)); \
  uint8_t* op_d = Caml_ba_data_val(Field(d,0)); \
  size_t   o1   = Long_val(Field(a,1)); \
  size_t   o2   = Long_val(Field(b,1)); \
  size_t   o3   = Long_val(Field(c,1)); \
  size_t   o4   = Long_val(Field(c,1))

value bigstringaf_simd_abs_i8(value a, value b) {
  get_ops2(a,b);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i r  = _mm256_abs_epi8(ra);
  _mm256_store_si256((__m256i*)(op_b + o2), r);

  return Val_unit;
}

value bigstringaf_simd_add_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_add_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_adds_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_adds_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_adds_u8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_adds_epu8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_and_si256(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_and_si256(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_andnot_si256(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_andnot_si256(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_avg_u8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_avg_epu8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_blend_i8(value a, value b, value c, value d) {
  get_ops4(a,b,c,d);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i rc = _mm256_load_si256((__m256i*)(op_c + o3));
  __m256i r  = _mm256_blendv_epi8(ra, rb, rc);
  _mm256_store_si256((__m256i*)(op_d + o4), r);

  return Val_unit;
}

value bigstringaf_simd_broadcast_i8(value a, value b) {
  get_ops1(a);
  uint8_t constant = (uint8_t)(Int_val(b));

  __m256i r = _mm256_set1_epi8(constant);
  _mm256_store_si256((__m256i*)(op_a + o1), r);

  return Val_unit;
}

value bigstringaf_simd_cmpeq_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r = _mm256_cmpeq_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_cmpgt_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r = _mm256_cmpgt_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_max_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r = _mm256_max_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_max_u8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r = _mm256_max_epu8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_min_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r = _mm256_min_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_min_u8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r = _mm256_min_epu8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_movemask_i8(value a) {
  get_ops1(a);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  uint32_t r  = _mm256_movemask_epi8(ra);

  return Val_long(r);
}

value bigstringaf_simd_or_si256(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_or_si256(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_set_i8(value a, value b) {
  get_ops1(a);

  __m256i r = _mm256_setr_epi8(
    (uint8_t)Int_val(Field(b,0)),
    (uint8_t)Int_val(Field(b,1)),
    (uint8_t)Int_val(Field(b,2)),
    (uint8_t)Int_val(Field(b,3)),
    (uint8_t)Int_val(Field(b,4)),
    (uint8_t)Int_val(Field(b,5)),
    (uint8_t)Int_val(Field(b,6)),
    (uint8_t)Int_val(Field(b,7)),
    (uint8_t)Int_val(Field(b,8)),
    (uint8_t)Int_val(Field(b,9)),
    (uint8_t)Int_val(Field(b,10)),
    (uint8_t)Int_val(Field(b,11)),
    (uint8_t)Int_val(Field(b,12)),
    (uint8_t)Int_val(Field(b,13)),
    (uint8_t)Int_val(Field(b,14)),
    (uint8_t)Int_val(Field(b,15)),
    (uint8_t)Int_val(Field(b,16)),
    (uint8_t)Int_val(Field(b,17)),
    (uint8_t)Int_val(Field(b,18)),
    (uint8_t)Int_val(Field(b,19)),
    (uint8_t)Int_val(Field(b,20)),
    (uint8_t)Int_val(Field(b,21)),
    (uint8_t)Int_val(Field(b,22)),
    (uint8_t)Int_val(Field(b,23)),
    (uint8_t)Int_val(Field(b,24)),
    (uint8_t)Int_val(Field(b,25)),
    (uint8_t)Int_val(Field(b,26)),
    (uint8_t)Int_val(Field(b,27)),
    (uint8_t)Int_val(Field(b,28)),
    (uint8_t)Int_val(Field(b,29)),
    (uint8_t)Int_val(Field(b,30)),
    (uint8_t)Int_val(Field(b,31)));

  _mm256_store_si256((__m256i*)(op_a + o1), r);

  return Val_unit;
}

value bigstringaf_simd_shuffle_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r = _mm256_shuffle_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_sign_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r = _mm256_sign_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_sub_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_sub_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_subs_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_subs_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_subs_u8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_subs_epu8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_testc_si256(value a, value b) {
  get_ops2(a,b);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  uint32_t r  = _mm256_testc_si256(ra, rb);

  return Val_long(r);
}

value bigstringaf_simd_testnzc_si256(value a, value b) {
  get_ops2(a,b);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  uint32_t r  = _mm256_testnzc_si256(ra, rb);

  return Val_long(r);
}

value bigstringaf_simd_testz_si256(value a, value b) {
  get_ops2(a,b);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  uint32_t r  = _mm256_testz_si256(ra, rb);

  return Val_long(r);
}

value bigstringaf_simd_unpackhi_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_unpackhi_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_unpacklo_i8(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_unpacklo_epi8(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}

value bigstringaf_simd_xor_si256(value a, value b, value c) {
  get_ops3(a,b,c);

  __m256i ra = _mm256_load_si256((__m256i*)(op_a + o1));
  __m256i rb = _mm256_load_si256((__m256i*)(op_b + o2));
  __m256i r  = _mm256_xor_si256(ra, rb);
  _mm256_store_si256((__m256i*)(op_c + o3), r);

  return Val_unit;
}
